dev-tools/ls-apis/src/workspaces.rs

```diff
@@ -0,0 +1,270 @@
+                Ok::<_, anyhow::Error>((workspace.name().to_owned(), workspace))
+            })
+            .collect::<Result<BTreeMap<_, _>, _>>()?;
+        workspaces.insert(
+            String::from("omicron"),
+            Arc::into_inner(omicron).expect("no more Omicron Arc references"),
+        );
+
+        // To load Dendrite, we need to look something up in Maghemite (loaded
+        // above).
+        let maghemite = workspaces
+            .get("maghemite")
+            .ok_or_else(|| anyhow!("missing maghemite workspaces"))?;
+
+        workspaces.insert(
+            String::from("dendrite"),
+            load_dependent_repo(&maghemite, "dendrite", "dpd-client")?,
+        );
+
+        // Validate the metadata against what we found in the workspaces.
+        let mut client_pkgnames_unused: BTreeSet<_> =
+            api_metadata.client_pkgnames().collect();
+        let mut warnings = Vec::new();
+        for (_, workspace) in &workspaces {
+            for client_pkgname in workspace.client_packages() {
+                if api_metadata.client_pkgname_lookup(client_pkgname).is_some()
+                {
+                    // It's possible that we will find multiple references
+                    // to the same client package name.  That's okay.
+                    client_pkgnames_unused.remove(client_pkgname);
+                } else {
+                    warnings.push(anyhow!(
+                        "workspace {}: found client package missing from API \
+                         manifest: {}",
+                        workspace.name(),
+                        client_pkgname
+                    ));
+                }
+            }
+        }
+
+        for c in client_pkgnames_unused {
+            warnings.push(anyhow!(
+                "API manifest refers to unknown client package: {}",
+                c
+            ));
+        }
+
+        Ok((Workspaces { workspaces }, warnings))
+    }
+
+    /// Given the name of a workspace package from one of our workspaces, return
+    /// the corresponding `Workspace` and `Package`
+    ///
+    /// This is only for finding packages defined *in* one of these workspaces.
+    /// For any other kind of package (e.g., transitive dependencies, which
+    /// might come from crates.io or other Git repositories), the name is not
+    /// unique and you'd need to use some other mechanism to get information
+    /// about it.
+    pub fn find_package_workspace(
+        &self,
+        server_pkgname: &str,
+    ) -> Result<(&Workspace, &Package)> {
+        // Figure out which workspace has this package.
+        let found_in_workspaces: Vec<_> = self
+            .workspaces
+            .values()
+            .filter_map(|w| {
+                w.find_workspace_package(&server_pkgname).map(|p| (w, p))
+            })
+            .collect();
+
+        // TODO As of this writing, we have two distinct packages called
+        // "dpd-client":
```

Comment:
Oh yeah this isn't good -- definitely worth fixing, maybe even as a blocker to this PR (which would make this comment moot). Probably easiest to rename the one in Omicron.

Reply:
I think this is not that easy because it may be used by dependencies that aren't in Omicron.  I will definitely file an issue for this.  I don't consider it a blocker because this is already a problem and not really made worse by having this tool.

Reply:
hmm, yeah that's unfortunate. Thanks for filing the issue.

Reply:
Filed https://github.com/oxidecomputer/omicron/issues/6666